File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/main/kotlin/com/runanywhere/sdk/foundation/errors/SDKException.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * SDKException — the canonical Exception wrapper around the proto-generated
 * SDKError message (ai.runanywhere.proto.v1.SDKError). All code throws
 * SDKException; the embedded proto SDKError carries the wire-canonical
 * payload (code, category, message, context, c_abi_code, nested_message).
 */

package com.runanywhere.sdk.foundation.errors

import ai.runanywhere.proto.v1.LogLevel
import com.runanywhere.sdk.infrastructure.logging.Logging
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge
import ai.runanywhere.proto.v1.ErrorCategory as ProtoErrorCategory
import ai.runanywhere.proto.v1.ErrorCode as ProtoErrorCode
import ai.runanywhere.proto.v1.ErrorContext as ProtoErrorContext
import ai.runanywhere.proto.v1.SDKError as ProtoSDKError

/**
 * SDKException — Exception subclass that wraps the proto-canonical SDKError.
 *
 * The embedded [error] field is the wire-canonical proto representation of
 * the error and contains:
 *  - `code` (proto ErrorCode, positive magnitude of C ABI code)
 *  - `category` (proto ErrorCategory, coarse routing bucket)
 *  - `message` (human-readable, non-localized)
 *  - `context` (optional source location + telemetry)
 *  - `c_abi_code` (optional negative `rac_result_t` integer for round-tripping)
 *  - `nested_message` (optional underlying-error message)
 *
 * @property error The proto-canonical SDKError wire payload
 */
class SDKException(
    val error: ProtoSDKError,
    cause: Throwable? = null,
) : Exception(error.message, cause) {
    /** The proto error code (positive magnitude of C ABI code). */
    val code: ProtoErrorCode get() = error.code

    /** The proto error category (coarse routing bucket). */
    val category: ProtoErrorCategory get() = error.category

    /** The negative `rac_result_t` integer from the C ABI, if available. */
    val cAbiCode: Int? get() = error.c_abi_code

    /** Optional underlying error message captured at wrap time. */
    val nestedMessage: String? get() = error.nested_message

    /** Stack trace captured by [Exception] at construction time. */
    val stackTraceSnapshot: List<String> = stackTrace.map { it.toString() }

    /** LocalizedError-style description parity with Swift. */
    val errorDescription: String? get() = error.message.ifBlank { null }

    /** One-line failure-reason summary suitable for logs and UI diagnostics. */
    val failureReason: String
        get() = "[${category.name}] ${code.name}"

    /** Recovery guidance for common actionable errors. */
    val recoverySuggestion: String?
        get() =
            when (code) {
                ProtoErrorCode.ERROR_CODE_NOT_INITIALIZED ->
                    "Initialize the component before using it."
                ProtoErrorCode.ERROR_CODE_MODEL_NOT_FOUND ->
                    "Ensure the model is downloaded and the path is correct."
                ProtoErrorCode.ERROR_CODE_NETWORK_UNAVAILABLE ->
                    "Check your internet connection and try again."
                ProtoErrorCode.ERROR_CODE_INSUFFICIENT_STORAGE ->
                    "Free up storage space and try again."
                ProtoErrorCode.ERROR_CODE_INSUFFICIENT_MEMORY ->
                    "Close other applications to free up memory."
                ProtoErrorCode.ERROR_CODE_MICROPHONE_PERMISSION_DENIED ->
                    "Grant microphone permission in Settings."
                ProtoErrorCode.ERROR_CODE_TIMEOUT ->
                    "Try again or check your connection."
                ProtoErrorCode.ERROR_CODE_INVALID_API_KEY ->
                    "Verify your API key is correct."
                ProtoErrorCode.ERROR_CODE_CANCELLED ->
                    null
                else ->
                    null
            }

    /** Telemetry-only properties (lightweight, safe to ship). */
    val telemetryProperties: Map<String, String>
        get() =
            mapOf(
                "error_code" to code.name,
                "error_category" to category.name,
                "error_message" to error.message,
            )

    /**
     * Dot-separated path to the field that triggered a validation failure
     * (e.g. `"STTOptions.sample_rate"`). Populated by the generated
     * `validate()` helpers under `commonMain/.../generated/convenience/`
     * so callers can programmatically identify the failing field without
     * parsing the human-readable message.
     *
     * Backed by the typed `error.context.field_path` proto field.
     */
    val fieldPath: String? get() =
        error.context?.field_path?.takeIf { it.isNotEmpty() }

    override fun toString(): String =
        "SDKException[$category] ${code.name}: ${error.message}"

    companion object {
        // Generic factory (Swift parity: SDKException.make(...))

        /**
         * Generic factory; auto-logs unexpected errors unless [shouldLog] is
         * false or the [code] is classified as expected (e.g.
         * `ERROR_CODE_CANCELLED`, `ERROR_CODE_STREAM_CANCELLED`).
         *
         * Mirrors Swift's `SDKException.make(code:message:category:underlying:shouldLog:)`.
         *
         * When [cAbiCode] is `null` (the common case), the `c_abi_code` proto
         * field is auto-computed from [code] per the round-trip contract in
         * `errors.proto:516-518` (`cAbiCode = -code.value` for codes ≤ 899).
         * Callers that already hold the underlying `rac_result_t` value
         * (e.g. JNI bridge surfaces) may pass it explicitly to preserve the
         * lossless C ABI round-trip; that value should agree with the
         * negation of [code] per the same contract.
         *
         * @param code     The proto error code (positive magnitude of C ABI code).
         * @param message  Human-readable, non-localized error message.
         * @param category Coarse routing bucket. Defaults to
         *                 [ProtoErrorCategory.ERROR_CATEGORY_COMPONENT].
         * @param cAbiCode Optional explicit `rac_result_t` override. Null →
         *                 auto-compute via the round-trip contract.
         * @param cause    Optional underlying [Throwable] cause.
         * @param shouldLog When true (default), routes the exception through
         *                  [Logging] unless the code is classified as expected.
         */
        fun make(
            code: ProtoErrorCode,
            message: String,
            category: ProtoErrorCategory = ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
            cAbiCode: Int? = null,
            cause: Throwable? = null,
            shouldLog: Boolean = true,
        ): SDKException {
            val ex = of(code = code, category = category, message = message, cAbiCodeOverride = cAbiCode, cause = cause)
            if (shouldLog && !code.isExpected) {
                ex.log()
            }
            return ex
        }

        // Initialization factories

        fun notInitialized(component: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_NOT_INITIALIZED,
                category = ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
                message = "$component is not initialized",
                cause = cause,
            )

        fun invalidConfiguration(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_INVALID_CONFIGURATION,
                category = ProtoErrorCategory.ERROR_CATEGORY_CONFIGURATION,
                message = message,
                cause = cause,
            )

        fun invalidApiKey(cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_INVALID_API_KEY,
                category = ProtoErrorCategory.ERROR_CATEGORY_AUTH,
                message = "Invalid API key",
                cause = cause,
            )

        fun invalidArgument(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_INVALID_ARGUMENT,
                category = ProtoErrorCategory.ERROR_CATEGORY_VALIDATION,
                message = message,
                cause = cause,
            )

        fun validationFailed(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_VALIDATION_FAILED,
                category = ProtoErrorCategory.ERROR_CATEGORY_VALIDATION,
                message = message,
                cause = cause,
            )

        /**
         * Validation failure with a structured `fieldPath` discriminant
         * (e.g. `"STTOptions.sample_rate"`). Mirrors the canonical shape
         * emitted by `idl/codegen/generate_*_convenience.py`: every SDK
         * throws `{ code, category, fieldPath, message }`.
         *
         * Mirrors Swift's identical factory at SDKException.swift:196-222
         * which auto-logs the exception via `ex.log()` when the code is
         * not classified as expected. Validation failures (proto code
         * `ERROR_CODE_INVALID_ARGUMENT`) are never expected, so this path
         * always emits an ERROR-level log entry — keeping Kotlin / Swift
         * telemetry symmetric for the same misconfigured input.
         */
        fun validationFailed(
            fieldPath: String,
            message: String,
            cause: Throwable? = null,
        ): SDKException {
            val code = ProtoErrorCode.ERROR_CODE_INVALID_ARGUMENT
            val ex =
                SDKException(
                    error =
                        ProtoSDKError(
                            code = code,
                            category = ProtoErrorCategory.ERROR_CATEGORY_VALIDATION,
                            message = message,
                            context =
                                ProtoErrorContext(
                                    field_path = fieldPath,
                                ),
                            c_abi_code = roundTripCAbiCode(code),
                            nested_message = cause?.message,
                        ),
                    cause = cause,
                )
            if (!code.isExpected) {
                ex.log()
            }
            return ex
        }

        // Model factories

        fun modelNotFound(modelId: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_MODEL_NOT_FOUND,
                category = ProtoErrorCategory.ERROR_CATEGORY_MODEL,
                message = "Model not found: $modelId",
                cause = cause,
            )

        fun modelNotLoaded(modelId: String? = null, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_MODEL_NOT_LOADED,
                category = ProtoErrorCategory.ERROR_CATEGORY_MODEL,
                message = if (modelId != null) "Model not loaded: $modelId" else "No model is loaded",
                cause = cause,
            )

        fun modelLoadFailed(modelId: String, reason: String? = null, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_MODEL_LOAD_FAILED,
                category = ProtoErrorCategory.ERROR_CATEGORY_MODEL,
                message = if (reason != null) "Failed to load model $modelId: $reason" else "Failed to load model: $modelId",
                cause = cause,
            )

        // Network factories

        fun networkError(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_NETWORK_ERROR,
                category = ProtoErrorCategory.ERROR_CATEGORY_NETWORK,
                message = message,
                cause = cause,
            )

        /** Swift parity: SDKException.timeout(_:) — NETWORK-bucketed timeout. */
        fun timeout(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_TIMEOUT,
                category = ProtoErrorCategory.ERROR_CATEGORY_NETWORK,
                message = message,
                cause = cause,
            )

        // Component factories

        fun invalidState(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_INVALID_STATE,
                category = ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
                message = message,
                cause = cause,
            )

        fun authenticationFailed(reason: String? = null, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_AUTHENTICATION_FAILED,
                category = ProtoErrorCategory.ERROR_CATEGORY_AUTH,
                message = if (reason != null) "Authentication failed: $reason" else "Authentication failed",
                cause = cause,
            )

        fun unauthorized(resource: String? = null, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_UNAUTHORIZED,
                category = ProtoErrorCategory.ERROR_CATEGORY_AUTH,
                message = if (resource != null) "Unauthorized access to: $resource" else "Unauthorized access",
                cause = cause,
            )

        fun notImplemented(feature: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_NOT_IMPLEMENTED,
                category = ProtoErrorCategory.ERROR_CATEGORY_INTERNAL,
                message = "Not implemented: $feature",
                cause = cause,
            )

        /**
         * Swift parity: `SDKException.cancelled(_:)` — INTERNAL-bucketed
         * user-initiated cancellation. `shouldLog=false` because
         * `ERROR_CODE_CANCELLED` is classified as expected
         * (see [ProtoErrorCode.isExpected]).
         */
        fun cancelled(message: String = "Operation cancelled", cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_CANCELLED,
                category = ProtoErrorCategory.ERROR_CATEGORY_INTERNAL,
                message = message,
                cause = cause,
            )

        // Modality factories (STT/TTS/LLM/VAD/VLM/VoiceAgent).
        // Per errors.proto:516-518, `c_abi_code` MUST equal `-int32(code)` for
        // codes ≤ 899; the chosen ProtoErrorCode drives cAbiCode via of()'s
        // round-trip computation — no hand-written cAbiCode literals.

        fun tts(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_GENERATION_FAILED,
                category = ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
                message = message,
                cause = cause,
            )

        fun vlm(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_PROCESSING_FAILED,
                category = ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
                message = message,
                cause = cause,
            )

        fun voiceAgent(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_GENERATION_FAILED,
                category = ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
                message = message,
                cause = cause,
            )

        // Category factories (storage/platform/model/operation)

        fun storage(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_INSUFFICIENT_STORAGE,
                category = ProtoErrorCategory.ERROR_CATEGORY_IO,
                message = message,
                cause = cause,
            )

        fun platform(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_ADAPTER_NOT_SET,
                category = ProtoErrorCategory.ERROR_CATEGORY_CONFIGURATION,
                message = message,
                cause = cause,
            )

        fun model(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_MODEL_NOT_LOADED,
                category = ProtoErrorCategory.ERROR_CATEGORY_MODEL,
                message = message,
                cause = cause,
            )

        fun operation(message: String, cause: Throwable? = null) =
            of(
                code = ProtoErrorCode.ERROR_CODE_GENERATION_FAILED,
                category = ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
                message = message,
                cause = cause,
            )

        // Error conversion (Swift parity: SDKException.from / fromONNXCode)

        /**
         * Convert any [Throwable] into an SDKException. If [error] is already
         * an SDKException, returns it unchanged. A null [error] becomes an
         * `ERROR_CODE_UNKNOWN` envelope tagged with [category] and the message
         * `"Unknown error"`. Otherwise wraps in an `ERROR_CODE_UNKNOWN`
         * envelope tagged with [category] and the underlying throwable's
         * message.
         *
         * Mirrors Swift's `SDKException.from(_:category:)` (SDKException.swift:250-270).
         * The two Swift overloads (non-null and nullable) collapse into a
         * single nullable Kotlin signature because they share a JVM
         * descriptor. The Swift implementation additionally maps
         * `NSURLErrorDomain` codes — Kotlin defers that to platform-specific
         * HTTP adapters (e.g. [com.runanywhere.sdk.foundation.bridge.HTTPClientAdapter]).
         */
        fun from(
            error: Throwable?,
            category: ProtoErrorCategory = ProtoErrorCategory.ERROR_CATEGORY_INTERNAL,
        ): SDKException {
            if (error == null) {
                return make(
                    code = ProtoErrorCode.ERROR_CODE_UNKNOWN,
                    message = "Unknown error",
                    category = category,
                )
            }
            if (error is SDKException) return error
            return make(
                code = ProtoErrorCode.ERROR_CODE_UNKNOWN,
                message = error.message ?: error.toString(),
                category = category,
                cause = error,
            )
        }

        /**
         * Map an ONNX Runtime C error code into an SDKException.
         * Mirrors Swift's `SDKException.fromONNXCode(_:)` at SDKException.swift:299-341.
         */
        fun fromONNXCode(code: Int): SDKException {
            val (raCode, message) =
                when (code) {
                    0 -> ProtoErrorCode.ERROR_CODE_UNKNOWN to "Unexpected success code passed to error handler"
                    -1 -> ProtoErrorCode.ERROR_CODE_INITIALIZATION_FAILED to "ONNX Runtime initialization failed"
                    -2 -> ProtoErrorCode.ERROR_CODE_MODEL_LOAD_FAILED to "Failed to load ONNX model"
                    -3 -> ProtoErrorCode.ERROR_CODE_GENERATION_FAILED to "ONNX inference failed"
                    -4 -> ProtoErrorCode.ERROR_CODE_INVALID_STATE to "Invalid ONNX handle"
                    -5 -> ProtoErrorCode.ERROR_CODE_INVALID_INPUT to "Invalid ONNX parameters"
                    -6 -> ProtoErrorCode.ERROR_CODE_INSUFFICIENT_MEMORY to "ONNX Runtime out of memory"
                    -7 -> ProtoErrorCode.ERROR_CODE_NOT_IMPLEMENTED to "ONNX feature not implemented"
                    -8 -> ProtoErrorCode.ERROR_CODE_CANCELLED to "ONNX operation cancelled"
                    -9 -> ProtoErrorCode.ERROR_CODE_TIMEOUT to "ONNX operation timed out"
                    -10 -> ProtoErrorCode.ERROR_CODE_STORAGE_ERROR to "ONNX IO error"
                    else -> ProtoErrorCode.ERROR_CODE_UNKNOWN to "ONNX error code: $code"
                }
            return make(code = raCode, message = message, category = ProtoErrorCategory.ERROR_CATEGORY_INTERNAL)
        }

        /**
         * Map a `rac_result_t` (signed C ABI error [code]) to an
         * [SDKException] via the canonical commons helper
         * `rac_result_to_proto_error` (JNI [RunAnywhereBridge.racResultToProtoError]),
         * deserializing the returned `SDKError` proto. Returns `null` for
         * `RAC_SUCCESS` (0).
         *
         * Mirrors Swift's `SDKException.from(rcResult:)`
         * (RASDKError+Helpers.swift) so the rac_result_t → proto translation
         * lives in one place (commons) across every SDK instead of being
         * re-mapped here. Falls back to an `ERROR_CODE_UNKNOWN` envelope when
         * the bridge yields no payload (e.g. commons built without protobuf,
         * or the native library is unavailable).
         */
        fun fromRACResult(code: Int): SDKException? {
            if (code == RunAnywhereBridge.RAC_SUCCESS) return null
            val bytes =
                try {
                    RunAnywhereBridge.racResultToProtoError(code)
                } catch (_: Throwable) {
                    null
                }
            val proto = bytes?.let { runCatching { ProtoSDKError.ADAPTER.decode(it) }.getOrNull() }
            return if (proto != null) {
                SDKException(error = proto)
            } else {
                make(
                    code = ProtoErrorCode.ERROR_CODE_UNKNOWN,
                    message = "Unknown error code: $code",
                    category = ProtoErrorCategory.ERROR_CATEGORY_INTERNAL,
                    cAbiCode = code,
                )
            }
        }

        /**
         * Throw an [SDKException] if [code] indicates failure (non-zero).
         * Mirrors Swift's `SDKException.throwIfError(_:)`.
         */
        fun throwIfError(code: Int) {
            fromRACResult(code)?.let { throw it }
        }

        /**
         * Compute the round-trip `c_abi_code` for a proto [code] per the
         * errors.proto contract (lines 516-518): for codes in `1..899`, the
         * cAbiCode MUST equal `-code.value`; for Web-only WASM codes (≥ 900)
         * there is no canonical C ABI value, so cAbiCode is unset.
         *
         * Mirrors Swift's init at SDKException.swift:48-51.
         */
        private fun roundTripCAbiCode(code: ProtoErrorCode): Int? {
            val raw = code.value
            return if (raw in 1..899) -raw else null
        }

        /**
         * Internal helper to construct an SDKException with all fields set.
         * The `c_abi_code` defaults to [roundTripCAbiCode] of [code] per the
         * proto's round-trip contract; pass [cAbiCodeOverride] only when the
         * caller already holds the underlying `rac_result_t` value (e.g. JNI
         * bridge surfaces returning a concrete failure code).
         */
        private fun of(
            code: ProtoErrorCode,
            category: ProtoErrorCategory,
            message: String,
            cAbiCodeOverride: Int? = null,
            cause: Throwable? = null,
        ): SDKException =
            SDKException(
                error =
                    ProtoSDKError(
                        code = code,
                        category = category,
                        message = message,
                        c_abi_code = cAbiCodeOverride ?: roundTripCAbiCode(code),
                        nested_message = cause?.message,
                    ),
                cause = cause,
            )
    }
}

// ProtoErrorCode classification helper (Swift parity: RAErrorCode.isExpected)

/**
 * Whether this proto error code represents an expected/routine outcome that
 * SHOULD NOT be logged as an error (e.g. user-initiated cancellation).
 *
 * Mirrors Swift's `RAErrorCode.isExpected` extension — returns true only for
 * `ERROR_CODE_CANCELLED` and `ERROR_CODE_STREAM_CANCELLED`.
 */
val ProtoErrorCode.isExpected: Boolean
    get() =
        when (this) {
            ProtoErrorCode.ERROR_CODE_CANCELLED,
            ProtoErrorCode.ERROR_CODE_STREAM_CANCELLED,
            -> true
            else -> false
        }

// SDKException convenience extensions (Swift parity)

/**
 * Log this exception to the central [Logging] service.
 *
 * Mirrors Swift's `SDKException.log(file:line:function:)`. The level is
 * downgraded to [LogLevel.LOG_LEVEL_INFO] for [ProtoErrorCode.ERROR_CODE_CANCELLED];
 * all other codes log at [LogLevel.LOG_LEVEL_ERROR]. Call sites should typically gate
 * with `!code.isExpected` (the [SDKException.Companion.make] factory does
 * this automatically).
 *
 * @param file     Source file (default empty — pass via call-site if available).
 * @param line     Source line (default 0).
 * @param function Source function (default empty).
 */
fun SDKException.log(
    file: String = "",
    line: Int = 0,
    function: String = "",
) {
    val level: LogLevel = if (this.code == ProtoErrorCode.ERROR_CODE_CANCELLED) LogLevel.LOG_LEVEL_INFO else LogLevel.LOG_LEVEL_ERROR
    val fileName = file.substringAfterLast('/')

    val metadata =
        buildMap<String, Any?> {
            put("error_code", this@log.code.name)
            put("error_category", this@log.category.name)
            if (fileName.isNotEmpty()) put("source_file", fileName)
            if (line > 0) put("source_line", line)
            if (function.isNotEmpty()) put("source_function", function)
            this@log.cause?.let { put("underlying_error", it.toString()) }
            put("failure_reason", this@log.failureReason)
        }

    Logging.log(
        level = level,
        category = this.category.name,
        message = this.message ?: this.code.name,
        metadata = metadata,
        file = if (fileName.isNotEmpty()) fileName else null,
        line = if (line > 0) line else null,
        function = if (function.isNotEmpty()) function else null,
    )
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Foundation/Errors/SDKException.swift
//
//  SDKException.swift
//  RunAnywhere
//
//  Canonical Swift error type. Wraps the generated proto `RASDKError`
//  (Sources/RunAnywhere/Generated/errors.pb.swift) so Swift `throws` works
//  while keeping the wire-canonical proto as the source of truth.
//
//  Usage:
//      throw SDKException.modelNotFound("whisper-base")
//      throw SDKException(code: .notInitialized, message: "STT model not loaded", category: .component)
//      do { ... } catch let ex as SDKException { print(ex.proto.message) }
//

import Foundation

// MARK: - SDKException

/// Canonical Swift-throwable error wrapping the generated proto `RASDKError`.
public struct SDKException: Error, LocalizedError, Sendable, CustomStringConvertible {

    /// The canonical proto-encoded error this exception wraps.
    public let proto: RASDKError

    /// Optional underlying Swift error (not part of the wire proto).
    public let underlying: (any Error)?

    /// Stack trace captured at construction time (debug aid; not on the wire).
    public let stackTrace: [String]

    public init(proto: RASDKError, underlying: (any Error)? = nil) {
        self.proto = proto
        self.underlying = underlying
        self.stackTrace = Thread.callStackSymbols
    }

    public init(
        code: RAErrorCode,
        message: String,
        category: RAErrorCategory = .component,
        underlying: (any Error)? = nil
    ) {
        var proto = RASDKError()
        proto.code = code
        proto.message = message
        proto.category = category
        // Round-trip C ABI code: positive proto code ↔ negative rac_result_t
        let raw = code.rawValue
        if raw > 0 && raw <= 899 {
            proto.cAbiCode = -Int32(raw)
        }
        if let unwrapped = underlying {
            proto.nestedMessage = String(describing: unwrapped)
        }
        self.proto = proto
        self.underlying = underlying
        self.stackTrace = Thread.callStackSymbols
    }

    // MARK: Convenience accessors

    public var code: RAErrorCode { proto.code }
    public var category: RAErrorCategory { proto.category }
    public var message: String { proto.message }

    /// Dot-separated path to the field that triggered a validation failure
    /// (e.g. `"STTOptions.sampleRate"`). Populated by the generated
    /// `validate()` helpers under `Sources/RunAnywhere/Generated/...` so
    /// callers can programmatically identify the failing field without
    /// parsing the human-readable message.
    ///
    /// Backed by the typed `context.fieldPath` proto field, which is the
    /// wire-canonical carrier shared with Kotlin / Dart / TS.
    public var fieldPath: String? {
        guard proto.hasContext, proto.context.hasFieldPath else { return nil }
        return proto.context.fieldPath
    }

    // MARK: LocalizedError

    public var errorDescription: String? { proto.message }

    public var failureReason: String? {
        "[\(proto.category)] \(proto.code)"
    }

    public var recoverySuggestion: String? {
        switch proto.code {
        case .notInitialized:
            return "Initialize the component before using it."
        case .modelNotFound:
            return "Ensure the model is downloaded and the path is correct."
        case .networkUnavailable:
            return "Check your internet connection and try again."
        case .insufficientStorage:
            return "Free up storage space and try again."
        case .insufficientMemory:
            return "Close other applications to free up memory."
        case .microphonePermissionDenied:
            return "Grant microphone permission in Settings."
        case .timeout:
            return "Try again or check your connection."
        case .invalidApiKey:
            return "Verify your API key is correct."
        case .cancelled:
            return nil
        default:
            return nil
        }
    }

    // MARK: CustomStringConvertible

    public var description: String {
        var result = "SDKException[\(proto.category).\(proto.code)]: \(proto.message)"
        if let unwrapped = underlying {
            result += "\n  Caused by: \(unwrapped)"
        }
        return result
    }

    /// Telemetry-only properties (lightweight, safe to ship).
    public var telemetryProperties: [String: String] {
        [
            "error_code": "\(proto.code)",
            "error_category": "\(proto.category)",
            "error_message": proto.message
        ]
    }
}

// MARK: - Equatable / Hashable

extension SDKException: Equatable {
    public static func == (lhs: SDKException, rhs: SDKException) -> Bool {
        lhs.proto.code == rhs.proto.code &&
        lhs.proto.category == rhs.proto.category &&
        lhs.proto.message == rhs.proto.message
    }
}

extension SDKException: Hashable {
    public func hash(into hasher: inout Hasher) {
        hasher.combine(proto.code.rawValue)
        hasher.combine(proto.category.rawValue)
        hasher.combine(proto.message)
    }
}

// MARK: - Generic factory

extension SDKException {
    /// Generic factory; auto-logs unexpected errors.
    public static func make(
        code: RAErrorCode,
        message: String,
        category: RAErrorCategory = .component,
        underlying: (any Error)? = nil,
        shouldLog: Bool = true
    ) -> SDKException {
        let ex = SDKException(code: code, message: message, category: category, underlying: underlying)
        if shouldLog && !code.isExpected {
            ex.log()
        }
        return ex
    }
}

// MARK: - Common shortcuts

extension SDKException {
    /// Common shortcut: model not found.
    public static func modelNotFound(_ id: String) -> SDKException {
        make(code: .modelNotFound, message: "Model not found: \(id)", category: .model)
    }

    /// Common shortcut: not initialized.
    public static func notInitialized(_ what: String) -> SDKException {
        make(code: .notInitialized, message: "\(what) is not initialized", category: .component)
    }

    /// Common shortcut: invalid configuration.
    public static func invalidConfiguration(_ message: String) -> SDKException {
        make(code: .invalidConfiguration, message: message, category: .configuration)
    }

    /// Common shortcut: validation failed.
    public static func validationFailed(_ message: String) -> SDKException {
        make(code: .validationFailed, message: message, category: .validation)
    }

    /// Validation failure with a structured `fieldPath` discriminant
    /// (e.g. `"STTOptions.sampleRate"`). Mirrors the canonical shape
    /// emitted by `idl/codegen/generate_*_convenience.py`: every SDK
    /// throws `{ code, category, fieldPath, message }`.
    public static func validationFailed(
        fieldPath: String,
        message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        var context = RAErrorContext()
        context.fieldPath = fieldPath

        var proto = RASDKError()
        proto.code = .invalidArgument
        proto.category = .validation
        proto.message = message
        proto.context = context
        // Round-trip C ABI code: positive proto code ↔ negative rac_result_t.
        let raw = RAErrorCode.invalidArgument.rawValue
        if raw > 0 && raw <= 899 {
            proto.cAbiCode = -Int32(raw)
        }
        if let unwrapped = underlying {
            proto.nestedMessage = String(describing: unwrapped)
        }
        let ex = SDKException(proto: proto, underlying: underlying)
        if !RAErrorCode.invalidArgument.isExpected {
            ex.log()
        }
        return ex
    }

    /// Common shortcut: cancelled.
    public static func cancelled(_ message: String = "Operation cancelled") -> SDKException {
        make(code: .cancelled, message: message, category: .internal, shouldLog: false)
    }

    /// Common shortcut: not implemented.
    public static func notImplemented(_ message: String) -> SDKException {
        make(code: .notImplemented, message: message, category: .internal)
    }

    /// Common shortcut: timeout.
    public static func timeout(_ message: String) -> SDKException {
        make(code: .timeout, message: message, category: .network)
    }

    /// Common shortcut: network error.
    public static func networkError(_ message: String) -> SDKException {
        make(code: .networkError, message: message, category: .network)
    }
}

// MARK: - Conversion from arbitrary Error

extension SDKException {
    /// Convert any `Error` into an `SDKException`. If already one, returns it
    /// unchanged. Otherwise wraps in an unknown / general error.
    public static func from(_ error: any Error, category: RAErrorCategory = .internal) -> SDKException {
        if let ex = error as? SDKException { return ex }

        let nsError = error as NSError
        if nsError.domain == NSURLErrorDomain {
            return fromURLError(nsError, category: category)
        }
        return make(
            code: .unknown,
            message: error.localizedDescription,
            category: category,
            underlying: error
        )
    }

    public static func from(_ error: (any Error)?, category: RAErrorCategory = .internal) -> SDKException {
        guard let error = error else {
            return make(code: .unknown, message: "Unknown error", category: category)
        }
        return from(error, category: category)
    }

    private static func fromURLError(_ nsError: NSError, category: RAErrorCategory) -> SDKException {
        let code: RAErrorCode
        switch nsError.code {
        case NSURLErrorNotConnectedToInternet, NSURLErrorNetworkConnectionLost:
            code = .networkUnavailable
        case NSURLErrorTimedOut:
            code = .timeout
        case NSURLErrorCancelled:
            code = .cancelled
        case NSURLErrorCannotFindHost, NSURLErrorCannotConnectToHost:
            code = .networkError
        default:
            code = .networkError
        }
        return make(
            code: code,
            message: nsError.localizedDescription,
            category: category,
            underlying: nsError
        )
    }
}

// MARK: - ONNX Runtime error mapping

extension SDKException {
    /// Map an ONNX Runtime C error code into an SDKException.
    public static func fromONNXCode(_ code: Int32) -> SDKException {
        let raCode: RAErrorCode
        let message: String
        switch code {
        case 0:
            raCode = .unknown
            message = "Unexpected success code passed to error handler"
        case -1:
            raCode = .initializationFailed
            message = "ONNX Runtime initialization failed"
        case -2:
            raCode = .modelLoadFailed
            message = "Failed to load ONNX model"
        case -3:
            raCode = .generationFailed
            message = "ONNX inference failed"
        case -4:
            raCode = .invalidState
            message = "Invalid ONNX handle"
        case -5:
            raCode = .invalidInput
            message = "Invalid ONNX parameters"
        case -6:
            raCode = .insufficientMemory
            message = "ONNX Runtime out of memory"
        case -7:
            raCode = .notImplemented
            message = "ONNX feature not implemented"
        case -8:
            raCode = .cancelled
            message = "ONNX operation cancelled"
        case -9:
            raCode = .timeout
            message = "ONNX operation timed out"
        case -10:
            raCode = .storageError
            message = "ONNX IO error"
        default:
            raCode = .unknown
            message = "ONNX error code: \(code)"
        }
        return make(code: raCode, message: message, category: .internal)
    }
}

// MARK: - RAErrorCode classification helper

extension RAErrorCode {
    /// Whether this error is expected/routine and shouldn't be logged as error.
    public var isExpected: Bool {
        switch self {
        case .cancelled, .streamCancelled:
            return true
        default:
            return false
        }
    }
}

// MARK: - Logging hook

extension SDKException {
    /// Log this exception to all configured destinations.
    public func log(file: String = #file, line: Int = #line, function: String = #function) {
        let level: RALogLevel = (proto.code == .cancelled) ? .info : .error
        let fileName = (file as NSString).lastPathComponent

        var metadata: [String: Any] = [ // swiftlint:disable:this prefer_concrete_types avoid_any_type
            "error_code": "\(proto.code)",
            "error_category": "\(proto.category)",
            "source_file": fileName,
            "source_line": line,
            "source_function": function
        ]

        if let underlying = underlying {
            metadata["underlying_error"] = String(describing: underlying)
        }
        if let reason = failureReason {
            metadata["failure_reason"] = reason
        }

        // Top SDK frames only (cheap and useful)
        let sdkFrames = stackTrace.filter { $0.contains("RunAnywhere") }.prefix(5)
        if !sdkFrames.isEmpty {
            metadata["stack_trace"] = sdkFrames.joined(separator: "\n")
        }

        Logging.shared.log(
            level: level,
            category: "\(proto.category)",
            message: proto.message,
            metadata: metadata
        )
    }
}

// MARK: - C ABI helpers

extension SDKException {
    /// Map a `rac_result_t` code to an SDKException, or nil on success.
    public var rawCABICode: Int32 {
        proto.cAbiCode
    }
}
